framework/src/mediaplayer/mediaplayer.c

```diff
@@ -0,0 +1,581 @@
+			case MEDIA_STATE_PLAYING:
+#ifndef CONFIG_AUDIO_MULTI_CARD
+				if (recording) {
+					node = dq_next(node);
+					continue;
+				}
+				if (!playing) {
+					pcmout = pcm_open(0, 0, PCM_OUT, &config);
+					buffer_size = pcm_get_buffer_size(pcmout);
+					playing = true;
+					buffer = (char *)malloc(buffer_size);
+					if (buffer == NULL) {
+						((media_t *)node)->state = MEDIA_STATE_STOPPING;
+						node = dq_next(node);
+						continue;
+					}
+				}
+
+#endif
+				ret = on_media_state_playing(pcmout, (media_t *)node, buffer, buffer_size);
+				if (ret == 0) {
+					// If playing is done,
+					((media_t *)node)->state = MEDIA_STATE_STOPPING;
+				} else if (ret < 0) {
+					// Todo: Error handling
+					printf("Playing error!\n");
+					((media_t *)node)->state = MEDIA_STATE_STOPPING;
+				}
+				break;
+#ifndef CONFIG_AUDIO_MULTI_CARD
+			case MEDIA_STATE_RECORDING:
+				if (playing) {
+					node = dq_next(node);
+					continue;
+				}
+				if (!recording) {
+					pcmin = pcm_open(0, 0, PCM_IN, &config);
+					buffer_size = pcm_get_buffer_size(pcmin);
+					buffer = (char *)malloc(buffer_size);
+					recording = true;
+					if (buffer == NULL) {
+						((media_t *)node)->state = MEDIA_STATE_STOPPING;
+						node = dq_next(node);
+						continue;
+					}
+				}
+				if (on_media_state_recording(pcmin, (media_t *)node, buffer, buffer_size) < 0) {
+					// Todo: Error handling
+					((media_t *)node)->state = MEDIA_STATE_STOPPING;
+					printf("Recording error!\n");
+				}
+
+				break;
+#endif
+
+			case MEDIA_STATE_PAUSING:
+				((media_t *)node)->state = MEDIA_STATE_PAUSED;
+				sem_wait(&g_sem_playing);
+				break;
+
+			case MEDIA_STATE_PAUSED:
+				break;
+
+			case MEDIA_STATE_STOPPING:
+#ifndef CONFIG_AUDIO_MULTI_CARD
+				if (playing) {
+					pcm_close(pcmout);
+					pcmout = NULL;
+					playing = false;
+				}
+				if (recording) {
+					pcm_close(pcmin);
+					pcmin = NULL;
+					recording = false;
+				}
+				if (buffer != NULL) {
+					free(buffer);
+				}
+#endif
+				dq_rem(node, &g_playing_q);
+				sem_wait(&g_sem_playing);
+				break;
+			default:
+				break;
+			}
+
+			node = dq_next(node);
+		}
+		pthread_mutex_unlock(&g_mutex_playing_q);
+
+		usleep(1);
+	}
+
+	sem_post(&g_sem_playing);
+
+	pcm_close(pcmout);
+	pcmout = NULL;
+#ifndef CONFIG_AUDIO_MULTI_CARD
+	pcm_close(pcmin);
+	pcmin = NULL;
+#endif
+
+	printf("audio_playing thread terminated.\n");
+	return MEDIA_OK;
+}
+
+media_result_t audio_recording(void *args)
+{
+	char *buffer;
+	unsigned int buffer_size;
+	struct pcm *pcmin;
+	struct pcm_config config;
+	dq_entry_t *node;
+
+	// Configuration
+	config.channels = 2;
+	config.rate = 16000;
+	config.period_size = CONFIG_AUDIO_BUFFER_NUMBYTES;
+	config.period_count = CONFIG_AUDIO_NUM_BUFFERS;
+	config.format = PCM_FORMAT_S16_LE;
+	config.start_threshold = 0;
+	config.stop_threshold = 0;
+	config.silence_threshold = 0;
+
+	pcmin = pcm_open(0, 0, PCM_IN, &config);
+	buffer_size = pcm_get_buffer_size(pcmin);
```

Comment:
If PR #717 added, this should be changed.

Reply:
yes it will change